Mathlib/Probability/ProbabilityMassFunction/Binomial.lean

```diff
@@ -0,0 +1,49 @@
+/-
+Copyright (c) 2023 Joachim Breitner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joachim Breitner
+-/
+import Mathlib.Probability.ProbabilityMassFunction.Constructions
+import Mathlib.Data.Nat.Choose.Multinomial
+import Mathlib.Tactic.FinCases
+
+/-!
+# The binomial distribution
+
+This file defines the probability mass function of the binomial distribution, and proves
+it to be equal to `Pmf.bernoulli` for `n = 1`.
+-/
+
+
+namespace Pmf
+
+/-- The binomial `Pmf`: The probability of that `i` out of `n` coins come up heads if the
+probability of heads is `p`. -/
+noncomputable
+def binomial (p : ENNReal) (h : p ≤ 1) (n : ℕ) : Pmf (Fin (n + 1)) :=
+  .ofFintype (fun i => p^(i : ℕ) * (1-p)^(n - (i : ℕ)) * (n.choose i : ℕ)) (by
```

Comment:
This looks like a job for `Finset.Nat.antidiagonal`:
```suggestion
def binomial (p : ENNReal) (h : p ≤ 1) (n : ℕ) : Pmf (antidiagonal n) :=
  .ofFintype (fun ⟨⟨i, j⟩, h⟩ ↦ p^i * (1-p)^j * ((i + j).choose j : ℕ)) (by
```

Reply:
I was actually thinking of providing both variants, just like most of the `choose` API has both, if I recall. As a user id certainly expect to have the binomial distribution on `Fin (n+1)` somewhere.

Reply:
I suggest we wait and see if there is demand for the `Fin (n + 1)`-version.

I admit there are good chances such demand may eventually arise (maybe even soon!) but I think we should at least try to avoid having duplicate definitions for the same thing.

Reply:
Hmm. I'm skeptical how it will look like in my project (<https://github.com/nomeata/rerolling-sixes/blob/master/RerollingSixes.lean>).

Do we have a function `Fintyp (n+1) -> antidiagonal n` so that I can easily map back and forth?

I'd was inclined to follow the example set by https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Choose/Sum.html#Commute.add_pow' and `add_pow`. But I can of course do the diagonal variant first.

Reply:
> I'd was inclined to follow the example set by https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Choose/Sum.html#Commute.add_pow' and `add_pow`. But I can of course do the diagonal variant first.

I don't think this is such a good analogy because `Commute.add_pow` and `Commute.add_pow'` are both theorems but this is a definition.

More subjectively, I claim `Commute.add_pow` is not a great lemma because it uses `Nat.sub`. The use of `Nat.sub` here is my main reason for pushing `Finset.Nat.antidiagonal` but it's not the only way to avoid `Nat.sub`. For example, bearing in mind:
```
example  (n : ℕ) (i : Fin (n + 1)) : (↑(Fin.last n - i) : ℕ) = n - (i : ℕ) := by simp [Fin.last_sub]
```
we could have a `Fin (n + 1)`-supported binomial distribution without using `Nat.sub` as follows:
```suggestion
def binomial (p : ENNReal) (h : p ≤ 1) (n : ℕ) : Pmf (Fin (n + 1)) :=
  .ofFintype (fun i => p^(i : ℕ) * (1-p)^(↑(Fin.last n - i) : ℕ) * (n.choose i : ℕ)) (by
```

Reply:
> Interesting, is `Fin.sub` less worthy of avoidance than `Nat.sub`? I understand that `Fin.sub` is annoying;
There's nothing wrong with `Fin.sub`, it's just modular arithmetic. Indeed:
```lean
example (n : ℕ) : AddCommGroup (Fin (n + 1)) := inferInstance
```
so it's a sensible subtraction in which things like `sub_eq_zero` etc. hold. 

>  I would hope that we’d build an API around binomial so that one doesn’t often have to drop down to its actual definition, so it matters a bit less what’s happening on the RHS of the `Pmf.binominal`
Agreed, but if we can avoid `Nat.sub` (as it seems we can) then we should.

> but it’s type (i.e. whether the carrier is `Nat` or `Fin` or `antidiagonal`) matters much more
Also agreed. Note that there should be API to make it trivial to transport pmfs between domains (and if not we should create it). Given that we can do whatever we like if we also add the following (which seems to be missing, but is worth adding):
```
def antidiagonal_equiv_fin (n : ℕ) : antidiagonal n ≃ Fin (n + 1) where
  toFun := fun ⟨⟨i, j⟩, h⟩ ↦ ⟨i, by
    simp only [mem_antidiagonal] at h
    rw [Nat.lt_succ]
    -- etc
    sorry⟩
  invFun := fun ⟨i, h⟩ ↦ ⟨⟨i, n - i⟩, by
    rw [mem_antidiagonal, add_comm, tsub_add_cancel_iff_le]
    rwa [Nat.lt_succ] at h⟩
  left_inv := by
    rintro ⟨⟨i, j⟩, h⟩
    simp only [ge_iff_le, Subtype.mk.injEq, Prod.mk.injEq, true_and]
    simp only [mem_antidiagonal] at h
    exact tsub_eq_of_eq_add_rev h.symm
  right_inv := fun x ↦ by simp
```

But I guess this is all moot if we just use the `Fin.sub` version.

Reply:
Ok, I'll use the `Fin (n+1)` and `Fin.sub` variant, that sounds like a good sweet spot, thanks!

I'll also PR that equivalence.